keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,577 @@
+on isolated networks. The control and cluster networks may have overlapping IP addresses. 
+There for they require a non IP routing layer (SSH tunnel are an example).
+Adding this layer enables metadata audit logging. It allows validation of outgoing API server connections.
+Structuring the API server in this way is a forcing function for keeping architectural layering violations out of apiserver.
+In combination with a firewall, this separation of networks protects against security concerns such as 
+[Security Impact of Kubernetes API server external IP address proxying](https://groups.google.com/d/msg/kubernetes-security-announce/tyd-MVR-tY4/tyREP9-qAwAJ).
+
+### Goals
+
+Delete the SSH Tunnel/Node Dialer code from Kube APIServer.
+Enable admins to fix https://groups.google.com/d/msg/kubernetes-security-announce/tyd-MVR-tY4/tyREP9-qAwAJ.
+Allow isolation of the Control network from the Cluster network.
+
+### Non-Goals
+
+Build a general purpose Proxy which does everything. (Users should build their own
+custom proxies with the desired behavior, based on the provided proxy)
+
+## Definitions
+
+- **Master Network** An IP reachable network space which contains the master components, such as Kubernetes API Server, 
+Connectivity Proxy and ETCD server.
+- **Cluster Network** An IP reachable network space which contains (some/all?) the clusters Nodes. 
+Worth noting that the Cluster Network may be completely disjoint from the Master network. 
+It may have overlapping IP addresses to the Master Network or other means of network isolation. 
+Direct IP routability between cluster and master networks should not be assumed.
+- **KAS** Kubernetes API Server, responsible for serving the Kubernetes API to clients.
+- **KMS** Key Management Service, plugins for secrets encryption key management
+- **Connectivity service** A component built into the KAS which provides a golang dialer for outgoing connection requests. 
+The dialer provided depends on NetworkContext information.
+- **Connectivity proxy** The proxy that runs in the master network. 
+It has a secure channel established to the cluster network. 
+It exposes a gRPC interface to KAS to provide connectivity service.
+- **Flat Network** A network where traffic can be successfully routed using IP. 
+Implies no overlapping (i.e. shared) IPs on the network.
+
+## Proposal
+
+We will run a connectivity proxy inside the master network. 
+The connectivity proxy exposes a gRPC interface to the KAS. 
+Each connectivity proxy allows secure connections to one or more cluster networks. 
+Any network addressed by a connectivity proxy must be flat. 
+Currently the only mechanism for handling overlapping IP ranges in Kubernetes is the Proxy. 
+Non IP routable traffic, past the proxy, would need to be a non Kubernetes mechanism to route.
+
+Running the connectivity proxy in a separate process has a few advantages. 
```

Comment:
small nit: I think this section would read better if it was in point form.

Reply:
Done